cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -11,3 +11,119 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	datapackagingv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+)
+
+// GetAvailablePackageSummaries returns the available packages managed by the 'kapp_controller' plugin
+func (s *Server) GetAvailablePackageSummaries(ctx context.Context, request *corev1.GetAvailablePackageSummariesRequest) (*corev1.GetAvailablePackageSummariesResponse, error) {
+	log.Infof("+kapp-controller GetAvailablePackageSummaries")
+
+	// Retrieve the proper parameters from the request
+	namespace := request.GetContext().GetNamespace()
+	cluster := request.GetContext().GetCluster()
+	pageSize := request.GetPaginationOptions().GetPageSize()
+	pageOffset, err := pageOffsetFromPageToken(request.GetPaginationOptions().GetPageToken())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to intepret page token %q: %v", request.GetPaginationOptions().GetPageToken(), err)
+	}
+	// Assume the default cluster if none is specified
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+	// fetch all the package metadatas
+	pkgMetadatas, err := s.getPkgMetadatas(ctx, cluster, namespace)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageMetadata", "", err)
+	}
+
+	// paginate the list of results
+	availablePackageSummaries := make([]*corev1.AvailablePackageSummary, len(pkgMetadatas))
+
+	// create the waiting group for processing each item aynchronously
+	var wg sync.WaitGroup
+
+	if len(pkgMetadatas) > 0 {
+		startAt := -1
+		if pageSize > 0 {
+			startAt = int(pageSize) * pageOffset
+		}
+		for i, pkgMetadata := range pkgMetadatas {
+			wg.Add(1)
+			if startAt <= i {
+				go func(i int, pkgMetadata *datapackagingv1alpha1.PackageMetadata) error {
+					defer wg.Done()
+					// fetch the associated packages
+					// Use the field selector to return only Package CRs that match on the spec.refName.
+					fieldSelector := fmt.Sprintf("spec.refName=%s", pkgMetadata.Name)
+					pkgs, err := s.getPkgsWithFieldSelector(ctx, cluster, namespace, fieldSelector)
```

Comment:
Makes me wonder, given that the package resources are pretty lean (no inlined SVG!), whether it'd be faster here to do a single request for all packages and just filter it down based on the metadata that we're dealing with. If you've got your cluster setup with the bitnami (carvel) repo, you can just try `time kubectl get packages --all-namespaces`.

That'd mean no need for go-routines here, if so.

Reply:
I assumed that delegating the pkg search thing back to kubernetes was quicker than doing it by ourselves, but you are possibly right... we better fetch the whole list -in memory: future issues??- and then we filter.

```
> time kubectl get packages --all-namespaces
real    0m1,464s
```

Reply:
Yeah, though if the inline SVG isn't the issue, my original thought above is moot.